mbtools/copy_resources_s3.py

```diff
@@ -0,0 +1,132 @@
+import hashlib
+import json
+import os
+import io
+import sys
+import magic
+import boto3
+import botocore
+from pathlib import Path
+
+
+def upload_resources(resource_dir, metadata_file, bucket, s3_dir):
+    """Uplaods resources to s3 if they dont already exist there"""
+
+    hash_to_filename_map = new_resource_hashes(resource_dir)
+    hashes_new = list(hash_to_filename_map.keys())
+
+    hashes_metadata = existing_metadata_hashes(metadata_file)
+    hashes_to_update = list(filter(lambda h: h not in hashes_metadata,
+                                   hashes_new)
+                            )
+    add_new_resources_to_s3(bucket, s3_dir, hashes_to_update,
+                            hash_to_filename_map, metadata_file
+                            )
+
+
+def existing_metadata_hashes(dir):
+    """Gathers the sha1s for existing resources from a local metadata json"""
+    hashes = []
+    with open(dir) as f:
+        data = json.load(f)
+        for item in data:
+            hashes.append(item["sha1"])
+    return hashes
+
+
+def new_resource_hashes(resource_dir):
+    """Generates sha1 hashes for all the resources in a local directory"""
+    sha1_map = {}
+    buff_size = io.DEFAULT_BUFFER_SIZE
+    for filename in os.listdir(resource_dir):
+
+        sha1 = hashlib.sha1()
+        full_path = os.path.join(resource_dir, filename)
+        with open(full_path, 'rb') as f:
+            while True:
+                data = f.read(buff_size)
+                if not data:
+                    break
+                sha1.update(data)
+        sha1_map[sha1.hexdigest()] = full_path
+    return sha1_map
+
+
+def get_mime_type(filename):
+    """Get the MIME type of file with libmagic """
+    mime_type = ''
+    try:
+        mime_type = magic.from_file(filename, mime=True)
+    finally:
+        return mime_type
+
+
+def add_new_resources_to_s3(bucket, s3_dir, hashes, hash2filename_map,
+                            metadata_file
+                            ):
+    """Add the files specified in filename_map to s3 iff they their
+    corresponding sha1 key exists in hashes"""
+
+    s3_client = boto3.client("s3")
+    hashkey2mimes_map = {}
+    hashkey2filenames_map = {}
+    hashkey2keypath_map = {}
+    for hash_key in hashes:
+        full_keypath = s3_dir + '/' + hash_key
+
+        try:
+            s3_client.head_object(Bucket=bucket,
+                                  Key=full_keypath)
```

Comment:
It may be nice to print a log message if a file is already found in S3 so the operator after this line.

Reply:
@rnathuji What specifically do you mean by a log file? Does mbtools have a log in system or is there one standard library that's easy to use... I think when I read this the first time I just thought print()

Reply:
@mjoss17 -- Oh, yeah, I just mean `print` for this. Maybe just something like:

```python
filename = os.path.basename(hash2filename_map[hash_key])
print(f"File {filename} with sha {hash_key} already exists in S3!")
```